runtime/src/bank/builtins/core_bpf_migration/mod.rs

```diff
@@ -86,14 +93,38 @@ impl Bank {
     /// Create an `AccountSharedData` with data initialized to
     /// `UpgradeableLoaderState::ProgramData` populated with the current slot, as
     /// well as the source program data account's upgrade authority and ELF.
+    ///
+    /// This function accepts a provided upgrade authority address, which comes
+    /// from the migration configuration. If the provided upgrade authority
+    /// address is different from the source buffer account's upgrade authority
+    /// address, the migration will fail. If the provided upgrade authority
+    /// address is `None`, the migration will ignore the source buffer account's
+    /// upgrade authority and set the new program data account's upgrade
+    /// authority to `None`.
     fn new_target_program_data_account(
         &self,
         source: &SourceBuffer,
+        upgrade_authority_address: Option<Pubkey>,
     ) -> Result<AccountSharedData, CoreBpfMigrationError> {
         let buffer_metadata_size = UpgradeableLoaderState::size_of_buffer_metadata();
-        if let UpgradeableLoaderState::Buffer { authority_address } =
-            bincode::deserialize(&source.buffer_account.data()[..buffer_metadata_size])?
+        if let UpgradeableLoaderState::Buffer {
+            authority_address: buffer_authority,
+        } = bincode::deserialize(&source.buffer_account.data()[..buffer_metadata_size])?
         {
+            let target_upgrade_authority_address =
+                match (upgrade_authority_address, buffer_authority) {
+                    (Some(provided_authority), Some(buffer_authority)) => {
+                        if provided_authority != buffer_authority {
+                            return Err(CoreBpfMigrationError::UpgradeAuthorityMismatch(
+                                provided_authority,
+                                buffer_authority,
+                            ));
+                        }
+                        Some(provided_authority)
+                    }
+                    _ => None,
+                };
```

Comment:
(Do remove my code comments)
```suggestion
            if upgrade_authority_address.is_some() {
                if upgrade_authority_address != buffer_authority {
                    return Err(CoreBpfMigrationError::UpgradeAuthorityMismatch(
                        upgrade_authority_address.unwrap(),
                        buffer_authority.unwrap_or_default(), // or switch this field to Option in the error definition
                    ));
                }
            }
            let target_upgrade_authority_address = upgrade_authority_address; // This line could be removed and upgrade_authority_address used directly; just being explicit and mirroring your code.
```

Reply:
This is cleaner but requires a `#[allow(clippy::unnecessary_unwrap)]`.